editor/src/clj/editor/gl/vertex2.clj

```diff
@@ -145,113 +96,89 @@
 (definline buf-put-floats! [buffer byte-offset numbers]
   `(buffers/put-floats! ~buffer ~byte-offset ~numbers))
 
-(definline buf-put! [buf byte-offset data-type normalize numbers]
-  `(buffers/put! ~buf ~byte-offset ~data-type ~normalize ~numbers))
-
-(defn put!
-  ^VertexBuffer [^VertexBuffer vbuf byte-offset data-type normalize numbers]
-  (buf-put! (.buf vbuf) byte-offset data-type normalize numbers)
-  vbuf)
+(definline buf-put! [buf byte-offset buffer-data-type normalize numbers]
+  `(buffers/put! ~buf ~byte-offset ~buffer-data-type ~normalize ~numbers))
 
 (definline buf-push-floats! [buffer numbers]
   `(buffers/push-floats! ~buffer ~numbers))
 
-(definline buf-push! [buf data-type normalize numbers]
-  `(buffers/push! ~buf ~data-type ~normalize ~numbers))
-
-(defn push!
-  ^VertexBuffer [^VertexBuffer vbuf data-type normalize numbers]
-  (buf-push! (.buf vbuf) data-type normalize numbers)
-  vbuf)
-
-;; vertex description
-
-(defn attribute-size
-  ^long [{:keys [^long components type]}]
-  (* components (buffers/type-size type)))
-
-(defn- vertex-size
-  ^long [attributes]
-  (transduce (map attribute-size) + 0 attributes))
-
-(defn make-vertex-description
-  [name attributes]
-  {:name name
-   :attributes attributes
-   :size (vertex-size attributes)})
-
+(definline buf-push! [buf buffer-data-type normalize numbers]
+  `(buffers/push! ~buf ~buffer-data-type ~normalize ~numbers))
 
 ;; defvertex macro
 
-(defn- attribute-components
-  [attributes]
-  (for [{:keys [components] :as attribute} attributes
-        n (range components)]
-    (-> attribute
-        (update :name str (nth ["-x" "-y" "-z" "-w"] n))
-        (dissoc :components))))
-
-(defn make-put-fn
-  [attributes]
-  (let [args (for [{:keys [name] :as component} (attribute-components attributes)]
+(defn- attribute-components [attribute-infos]
+  (for [{attribute-name :name :keys [name-key vector-type] :as attribute-info} attribute-infos
+        n (range (graphics.types/vector-type-component-count vector-type))]
+    (let [suffix (nth ["-x" "-y" "-z" "-w"] n)
+          attribute-name (str attribute-name suffix)
+          attribute-key (keyword (str (name name-key) suffix))]
+      (assoc attribute-info
+        :name attribute-name
+        :name-key attribute-key
+        :vector-type :vector-type-scalar))))
```

Comment:
We used to have many different representations of vertex attribute descriptions. Now, we return to return something that conforms to `graphics.types/attribute-info?` everywhere.